tests/integration/test_integration.py

```diff
@@ -161,21 +162,70 @@ async def _deploy_nrf(ops_test: OpsTest):
-
-
-async def _deploy_webui(ops_test: OpsTest):
+    await ops_test.model.integrate(relation1=f"{NRF_CHARM_NAME}:sdcore_config", relation2=NMS_MOCK)
+
+
+async def _deploy_nms_mock(ops_test: OpsTest):
+    fiveg_core_gnb_lib_url = "https://github.com/canonical/sdcore-nms-k8s-operator/raw/main/lib/charms/sdcore_nms_k8s/v0/fiveg_core_gnb.py"
+    fiveg_core_gnb_lib = requests.get(fiveg_core_gnb_lib_url, timeout=10).text
+    sdcore_config_lib_url = "https://github.com/canonical/sdcore-nms-k8s-operator/raw/main/lib/charms/sdcore_nms_k8s/v0/sdcore_config.py"
+    sdcore_config_lib = requests.get(sdcore_config_lib_url, timeout=10).text
+    any_charm_src_overwrite = {
+        "fiveg_core_gnb.py": fiveg_core_gnb_lib,
+        "sdcore_config.py": sdcore_config_lib,
+        "any_charm.py": textwrap.dedent(
+            """\
+        from fiveg_core_gnb import FivegCoreGnbProvides, PLMNConfig
+        from sdcore_config import SdcoreConfigProvides
+        from any_charm_base import AnyCharmBase
+        from ops.framework import EventBase
+        SDCORE_CONFIG_RELATION_NAME = "provide-sdcore-config"
+        CORE_GNB_RELATION_NAME = "provide-fiveg-core-gnb"
+
+        class AnyCharm(AnyCharmBase):
+            def __init__(self, *args, **kwargs):
+                super().__init__(*args, **kwargs)
+                self._sdcore_config = SdcoreConfigProvides(self, SDCORE_CONFIG_RELATION_NAME)
+                self._fiveg_core_gnb_provider = FivegCoreGnbProvides(self, CORE_GNB_RELATION_NAME)
+                self.framework.observe(
+                    self.on[SDCORE_CONFIG_RELATION_NAME].relation_changed,
+                    self.sdcore_config_relation_changed,
+                )
+                self.framework.observe(
+                    self.on[CORE_GNB_RELATION_NAME].relation_changed,
+                    self.fiveg_core_gnb_relation_changed,
+                )
+
+            def sdcore_config_relation_changed(self, event: EventBase) -> None:
+                sdcore_config_relations = self.model.relations.get(SDCORE_CONFIG_RELATION_NAME)
+                if not sdcore_config_relations:
+                    logger.info("No %s relations found.", SDCORE_CONFIG_RELATION_NAME)
+                    return
+                self._sdcore_config.set_webui_url_in_all_relations(webui_url="sdcore-nms:9876")
+
+            def fiveg_core_gnb_relation_changed(self, event: EventBase):
+                core_gnb_relations = self.model.relations.get(CORE_GNB_RELATION_NAME)
+                if not core_gnb_relations:
+                    logger.info("No %s relations found.", CORE_GNB_RELATION_NAME)
+                    return
+                for relation in core_gnb_relations:
+                    self._fiveg_core_gnb_provider.publish_gnb_config_information(
+                        relation_id=relation.id,
+                        tac=1,
+                        plmns=[PLMNConfig(mcc="001", mnc="01", sst=1, sd=1056816)],
+                    )
+        """
```

Comment:
I think this would be better in a separate file and reading it in for readability.

Reply:
This is moved to another file to increase readability.